lib/private/DB/QueryBuilder/Partitioned/PartitionedQueryBuilder.php

```diff
@@ -0,0 +1,425 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+namespace OC\DB\QueryBuilder\Partitioned;
+
+use OC\DB\QueryBuilder\CompositeExpression;
+use OC\DB\QueryBuilder\QuoteHelper;
+use OC\DB\QueryBuilder\Sharded\AutoIncrementHandler;
+use OC\DB\QueryBuilder\Sharded\ShardConnectionManager;
+use OC\DB\QueryBuilder\Sharded\ShardedQueryBuilder;
+use OCP\DB\IResult;
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\DB\QueryBuilder\IQueryFunction;
+use OCP\IDBConnection;
+
+/**
+ * A special query builder that automatically splits queries that span across multiple database partitions[1].
+ *
+ * This is done by inspecting the query as it's being built, and when a cross-partition join is detected,
+ * the part of the query that touches the partition is split of into a different sub-query.
+ * Then, when the query is executed, the results from the sub-queries are automatically merged.
+ *
+ * This whole process is intended to be transparent to any code using the query builder, however it does impose some extra
+ * limitation for queries that work cross-partition. See the documentation from `InvalidPartitionedQueryException` for more details.
+ *
+ * When a join is created in the query, this builder checks if it belongs to the same partition as the table from the
+ * original FROM/UPDATE/DELETE/INSERT and if not, creates a new "sub query" for the partition.
+ * Then for every part that is added the query, the part is analyzed to determine which partition the query part is referencing
+ * and the query part is added to the sub query for that partition.
+ *
+ * [1]: A set of tables which can't be queried together with the rest of the tables, such as when sharding is used.
+ */
+class PartitionedQueryBuilder extends ShardedQueryBuilder {
+	/** @var array<string, PartitionQuery> $splitQueries */
+	private array $splitQueries = [];
+	/** @var list<PartitionSplit> */
+	private array $partitions = [];
+
+	/** @var array{'select': string|array, 'alias': ?string}[] */
+	private array $selects = [];
+	private ?PartitionSplit $mainPartition = null;
+	private bool $hasPositionalParameter = false;
+	private QuoteHelper $quoteHelper;
+	private ?int $limit = null;
+	private ?int $offset = null;
+
+	public function __construct(
+		IQueryBuilder          $builder,
+		array                  $shardDefinitions,
+		ShardConnectionManager $shardConnectionManager,
+		AutoIncrementHandler   $autoIncrementHandler,
+	) {
+		parent::__construct($builder, $shardDefinitions, $shardConnectionManager, $autoIncrementHandler);
+		$this->quoteHelper = new QuoteHelper();
+	}
+
+	private function newQuery(): IQueryBuilder {
+		$builder = $this->builder->getConnection()->getQueryBuilder();
+		if ($builder instanceof PartitionedQueryBuilder) {
+			$builder = $builder->builder;
+		}
```

Comment:
How is that different from:
```suggestion
		$builder = $this->builder;
```

Reply:
It needs a fresh builder (added a comment)